### PR TITLE
Utility to verify compaction + fixes for bugs found

### PR DESCRIPTION
**Pull request by vgkholla**
This commit adds a utility that can verify compaction by using the compaction log and snapshots of the store before and after compaction.

It also fixes a few bugs that were found as a part of testing.

### CODE REVIEW

File: ambry-tools/src/main/java/com.github.ambry/store/CompactionVerifier.java [hunk 1]
```diff
@@ -0,0 +1,620 @@
...
+            "Compacted log should not contain segment: " + srcSegmentName;
+      } else {
+        assert tgtSegmentNames.contains(srcSegmentName) : "Compacted log is missing segment: " + srcSegmentName;
+        expectedTgtSegmentNames.add(srcSegmentName);
+      }
+    }
+
+    // 2. Check that log segments that are newly created have the right generation number
+    for (String segmentCompactedName : segmentsCompactedNames) {
+      String tgtSegmentName = LogSegmentNameHelper.getNextGenerationName(segmentCompactedName);
```
Comment (pnarayanan) [inline ambry-tools/src/main/java/com.github.ambry/store/CompactionVerifier.java]: Does this utility expect that compaction ran only once? Is that a reasonable?

Comment (vgkholla) [inline ambry-tools/src/main/java/com.github.ambry/store/CompactionVerifier.java]: There is no such expectation. Can clarify this offline

Also tested this out when a store is recompacted (twice)

File: ambry-tools/src/main/java/com.github.ambry/store/CompactionVerifier.java [hunk 2]
```diff
@@ -326,15 +335,30 @@ private void verifyStructure() {
...
+      if (tempCleanupFileExists) {
+        LOGGER.warn("A temp clean shutdown file exists");
+      }
+      if (filesOnDiskCount != inMemIndexSegmentCount) {
+        LOGGER.warn("There are {} index segment files on disk but {} in the in-mem index", filesOnDiskCount,
+            inMemIndexSegmentCount);
+      }
+    } else {
+      assert numTempFilesOnDisk == 0 : "Some log segments haven't been cleaned";
+      assert !tempCleanupFileExists : "The temp clean shutdown file has not been" + " deleted";
```
Comment (pnarayanan) [inline ambry-tools/src/main/java/com.github.ambry/store/CompactionVerifier.java]: unnecessary concatenation.

**Comment by vgkholla** [comment]
@pnarayanan Updated. Had to rebase. New changes since you approved are in 4fe6c6a and 	f8ba0d4. (and maybe e6289c5)

**Pull request merged**